fvm/state/state_holder.go

```diff
@@ -16,29 +30,226 @@ import (
+	}, nil
+}
+
+func (s *StateHolder) mergeIntoParent(
+	enforceLimits bool,
+) error {
+	if len(s.nestedTransactions) < 2 {
+		return fmt.Errorf("cannot commit the main transaction")
 	}
+
+	child := s.current()
+	s.nestedTransactions = s.nestedTransactions[:len(s.nestedTransactions)-1]
+	parent := s.current()
+
+	return parent.state.MergeState(child.state, enforceLimits)
 }
 
-// TODO(patrick): remove
-// State returns the active state
-func (s *StateHolder) State() *State {
-	return s.activeState
+// CommitToParent commits the changes in the current unrestricted nested
+// transaction to the parent (nested) transaction.
+func (s *StateHolder) CommitToParent() error {
+	if s.IsParseRestricted() {
+		// This is due to a programming error.
+		return fmt.Errorf(
+			"cannot commit unexpected/restricted nested transaction",
+		)
+	}
+
+	return s.mergeIntoParent(s.EnforceInteractionLimits())
 }
 
-// TODO(patrick): remove
-// SetActiveState sets active state
-func (s *StateHolder) SetActiveState(st *State) {
-	s.activeState = st
+// CommitToMain ensures there's only a single outstanding unrestricted nested
+// transaction and commits the changes to the main transaction.
+func (s *StateHolder) CommitToMain() error {
+	if s.NumNestedTransactions() > 1 {
+		return fmt.Errorf("found too many nested transactions")
+	}
+
+	return s.CommitToParent()
+}
+
+// CommitParseRestrictedToParent commits the changes in the current restricted
+// nested transaction to the parent (nested) transaction.  This returns
+// error if the specified location does not match the tracked location.
+func (s *StateHolder) CommitParseRestrictedToParent(
+	location common.AddressLocation,
+) (
+	*State,
+	error,
+) {
+	currentFrame := s.current()
+	if currentFrame.parseRestriction == nil ||
+		currentFrame.parseRestriction.ID() != location.ID() {
+
+		// This is due to a programming error.
+		return nil, fmt.Errorf(
+			"cannot commit unexpected nested transaction %v != %v",
+			currentFrame.parseRestriction,
+			location,
+		)
+	}
+
+	err := s.mergeIntoParent(s.EnforceInteractionLimits())
+	if err != nil {
+		return nil, err
+	}
+
+	return currentFrame.state, nil
+}
+
+// AttachAndCommitNestedTransactionForParsingProgram commits the changes in the
+// cached nested transaction to the current (nested) transaction.
+func (s *StateHolder) AttachAndCommitParseRestrictedToParent(
+	cachedNestedTransaction *State,
+) error {
+	s.nestedTransactions = append(
+		s.nestedTransactions,
+		nestedTransactionStackFrame{
+			state:            cachedNestedTransaction,
+			parseRestriction: nil,
+		},
+	)
+
+	// NOTE: limit enforcement is disabled here because cadence environment's
+	// Get() does not support returning error.
+	return s.mergeIntoParent(false)
+}
+
+// RestartNestedTransaction merges all changes that belongs to the nested
+// transaction about to be restart (for spock/meter bookkeeping), then
+// wipes its view changes.
+func (s *StateHolder) RestartNestedTransaction(
```

Comment:
Is this for optimisation purposes? If not I think it would be cleaner/more explicit to not have this at all. so you have to do commit and new.

Reply:
This is not an optimization.  This replaces/standardizes the old program handler's Cleanup method.

Restarting is needed to preserve the spock/meter bookkeeping.  (If we didn't have to worry about spock/meter, we can simply abort the nested transaction by dropping the stack frames, and create a new nested transaction)